File: groking_algorithm/binary_search/binary_search_recursive.js
function binary_search(list, k, low, high) {
  if (low > high)
    return null;
  else {
    mid = Math.floor((low+high) / 2);
    guess = list[mid].key;
    if (k === guess)
      return mid;
    if (k < guess)
      return binary_search(list, k, low, mid-1);
    else
      return binary_search(list, k, mid+1, high);
  }
}

const my_list = [
  { key: 1, val: "red" },
  { key: 3, val: "green" },
  { key: 5, val: "blue" },
  { key: 7, val: "yellow" }];

let item = -1;
let idx = binary_search(my_list, item, 0, my_list.length-1)
if (idx)
  console.log(`The key ${item} has value: ${my_list[idx].val}`);
else
  console.log(`The key is not found!`);

File: groking_algorithm/binary_search/binary_search_recursive.py
# Recursive version for binary search:
#   L: list
#   k: item to find
#   look for k for index range [low, high]
def binary_search(L, k, low, high):
  if low > high:
    return None
  else:
    mid = (low+high) // 2
    guess = L[mid]
    if k == guess:
      return mid
    if k < guess: 
      return binary_search(L, k, low, mid-1)
    else:
      return binary_search(L, k, mid+1, high)

my_list = [1, 3, 5, 7, 9]

print(binary_search(my_list, 3, 0, len(my_list)-1))
print(binary_search(my_list, -1, 0, len(my_list)-1))
print(binary_search(my_list, 7, 0, len(my_list)-1))